Ray step3

torch-ray opened the pull request:
## Step1 작업목록
- [x] 상품 이미지 추가
- [x] 재고, 잔액 Label 추가
- [x] 재고, 잔액 Button 추가
- [x] Button 누를 때 Label증가 코드 구현
- [x] IBOutlet Collection 학습
- [x] UIImage View radiusCorner 구현

## 학습키워드
- IBOutlet Collection
- MVC패턴

## 고민과 해결
- MVC패턴에 대해 이론적 학습 추가로 진행이 필요할 것 같습니다. 
- 코드로 구현하는 것에 있어서 V와 C를 구분하는 것이 어렵습니다.
- ViewController는 컨트롤러이고 View는 실행시에 나타나는 화면이라면, Model의 데이터를 활용하여 Viewcontroller에서 구현하면 결국 뷰와 컨트롤러를 분리한것 같습니다.
- 정리하자면 아직 ViewController에 어떻게 코드를 구현해야 MVC패턴에 적합한지 이해가어렵습니다.

## 부연설명
- (생략)

## 질문거리
- 뷰컨트롤러가 명확히 컨트롤러라면 Model에서 데이터를 가져와 뷰(실행시 나타나는 화면)를 보여주기 위해 코드를 구현하면되는 공간인 것 같습니다. 혹시 제 코드 중에 뷰컨트롤러에 있으면 좋지 않은(?) 코드는 무엇일까요? 버튼을 누르면 machine에 음료를 추가하는 건 어쩔 수없는 것 같고... 버튼에 따라 Label에 나타나는 재고의 양을 계산하는 것은 완전히 모델로 이전해야할까요? 이해 및 구현이 쉽지않아서ㅠㅠ 질문드립니다!

torch-ray committed:
이미지 추가 및 main.storyboard에 스택뷰 및 이미지 뷰 추가

torch-ray committed:
음료 이미지 삽입 및 스택뷰 + Label + Button 추가

torch-ray committed:
수량 label 및 재고추가 버튼 기능 구현

torch-ray committed:
프로퍼티명 변경 및 클로저 메서드 수정

Beverage class의 temparatureOfBeverage 프로퍼티 명을 temparature로 수정 및 Vedingmachine 구조체의 nowAvailableList메서드의 클로저를 Beverages로 이동하고nowAvailableList에서는 filterAvaiable을 호출하여 확인하도록 수정

torch-ray committed:
버튼 눌러서 음료 재고추가하는 기능 구현 완료

torch-ray committed:
잔액 뷰 추가 및 이미지뷰 roundCorner추가

torch-ray committed:
ReadMe.md 파일 추가

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -61,7 +61,7 @@ extension Beverage: CheckBeverageInfo {
     
     func isHotBeverage() -> Bool {
-        return temparatureOfBeverage >= 100 ? true : false
+        return temparature >= 100 ? true : false
```
- 이런 경우 상수로 선언하더라도 기준값을 매개변수로 전달하는 게 좋습니다. 
- 비교값인 경우는 `return 비교식` 정도로만 표현해도 충분합니다. 더 읽기 쉽구요.

torch-ray replied:
수정하였습니다. 추후 코딩에서도 참고하여 작업진행하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverages.swift:
```diff
@@ -31,4 +31,15 @@ class Beverages {
+                availableList.append(beverage)
+            }
+        }
+        return availableList
```
오잉? 자판기를 매번 만드는 건가요?

torch-ray replied:
불필요한 메서드라 삭제하고 VendingMachine에서 메서드 구현했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Coffee.swift:
```diff
@@ -4,9 +4,9 @@ class Coffee:Beverage {
     
     private var decaffeinated: Bool
```
여전히 Bool 타입으로 남아있는 속성이 있네요.

torch-ray replied:
디카페인 여부는 true/false로 구분할 수 있지 않을까요?!

godrm replied:
관점에 따라서 다를 수도 있을텐데요, 실제로도 caffeine 함량이 10mg 이하인 경우를 디카페인으로 분류한다고 합니다.
데이터를 모델링할 때 세부 속성은 더 구체적인 값을 갖고 있고, 일반적인 기준에 맞춰서 판단하는 것은 메소드로 제공하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -84,3 +71,8 @@ struct VendingMachine {
+extension VendingMachine: CustomStringConvertible {
+    var description: String {
+        return "\(insertedMoney)"
+    }
```
이 부분이 어떻게 활용되는 지 모르겠지만, 출력을 하기 위해서 억지로 변환하고 있는 것 같은 느낌입니다. 
맞나요? description은 보조적인 수단으로 사용하고, 로직에서, 출력용으로 사용하지 않는 게 좋습니다.

torch-ray replied:
삭제하였습니다!

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -2,9 +2,108 @@ import UIKit
     
+    private var machine = VendingMachine()
+    
+    @IBOutlet weak var stockInfoOfChocolateMilk: UILabel!
+    @IBOutlet weak var stockInfoOfStrawBerryMilk: UILabel!
+    @IBOutlet weak var stockInfoOfCokeZero: UILabel!
+    @IBOutlet weak var stockInfoOfSprite: UILabel!
+    @IBOutlet weak var stockInfoOfTOP: UILabel!
+    @IBOutlet weak var stockInfoOfCantata: UILabel!
+    @IBOutlet weak var balanceInfo: UILabel!
```
Outlet Collection에 대해서 학습해보세요. 혹시 개선할 수 있는지도 찾아보세요.

torch-ray replied:
Image view의 round coner는 동일한처리가 가능해서 collection을 사용하였는데, label 이나 button의 경우는 동일한 작업이 아니라 각자 처리가 필요해서 구현이 어려운 것 같습니다. 버튼의  tag와 label의 tag를 활용하면 될 것 같기도한데... 아직 수정이 어려운 것 같습니다. 혹시 추가 힌트를 주시면 방법을 찾아보겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -2,9 +2,108 @@ import UIKit
+extension VendingMachineViewController {
+    
+    private func beveragesStockCount() {
+        let stockOfBeverages = machine.beverages.countOfBeverages()
+        var (chocoMilke, strawberryMilk) = (0, 0)
+        var (coke, sprite) = (0, 0)
+        var (top, cantata) = (0, 0)
+        
+        for beverage in stockOfBeverages {
+            if beverage.key == "초코에몽" { chocoMilke += beverage.value }
+            else if beverage.key == "딸기꿀단지" { strawberryMilk += beverage.value }
+            else if beverage.key == "코카콜라제로" { coke += beverage.value }
+            else if beverage.key == "스프라이트" { sprite += beverage.value }
+            else if beverage.key == "TOP" { top += beverage.value }
+            else if beverage.key == "칸타타" { cantata += beverage.value }
+        }
+        
+        stockInfoOfChocolateMilk.text = "\(chocoMilke)개"
+        stockInfoOfStrawBerryMilk.text = "\(strawberryMilk)개"
+        stockInfoOfCokeZero.text = "\(coke)개"
+        stockInfoOfSprite.text = "\(sprite)개"
+        stockInfoOfTOP.text = "\(top)개"
+        stockInfoOfCantata.text = "\(cantata)개"
```
이건 좀 이상합니다. machine.beverages.countOfBeverages() 가 디미터 법칙을 지키지 않는 것 같습니다. 
machine.xxxx() 형태가 되어도 될 것 같구요. 
countOfBeverages() 메소드 이름만 보면 count만 넘겨주는 걸까요? Dictionary형태인가요? 
거기서 이미 구분해서 넘겨줬으면 if-else로 키를 비교하는 게 아니라, 각각 타입에 해당하는 label과 내용을 매칭하는 방법을 고려해보세요.
음료 종류별로 + 하고 있는 것은 자판기 하위에서 해야할 일 같습니다. 여기서는 받은 값을 text에 출력만 해야죠.

torch-ray replied:
딕셔너리를 [String:Int] 에서 [Beverage:Int]로 수정하고 음료별 카운트를 VendingMachine으로 메서드 이동하였습니다! 어떻게든 수정은 한 것 같은데 추가로 개선할 방법이 있을 것 같습니다 ㅠㅠ 조금만 더 힌트를 주시면 추후단계에서 수정하고 진행하겠습니다.

godrm replied:
네. 오늘 수업시간에 type(of:)와 ObjectIdentifier를 활용해도 좋겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -2,9 +2,108 @@ import UIKit
+    }
+    
+    @IBAction func buttonForChocolateMilkStock(_ sender: Any) {
+        machine.addBeverage(beverage: Beverage(brand: .namyang, volume: 200, price: 1000, productName: "초코에몽", manufacturedDay: Date(), sellByDate: Date(), calories: 100, temparature: 0))
```
이렇게 하드 코딩해서 Beverage를 생성하기 보다는 팩토리 객체에서 생성하는 것도 방법입니다.

torch-ray replied:
팩토리 객체를 생성했습니다! 맞는 방법인지는 모르겠습니다...ㅎㅎ

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -2,9 +2,108 @@ import UIKit
 
+extension VendingMachineViewController {
+    
+    private func balanceInfoLabel() {
+        let balance = machine.insertedMoney
+        balanceInfo.text = "잔액:\(balance)"
+    }
+    
+    @IBAction func button1000ForBalance(_ sender: Any) {
+        machine.getTheMoney(from: 1000)
+        balanceInfoLabel()
+    }
+    
+    @IBAction func button5000ForBalance(_ sender: Any) {
+        machine.getTheMoney(from: 5000)
+        balanceInfoLabel()
+    }
+    
```
이렇게 extension도 의도가 있을텐데 어떤 의도인지 구분하기 어렵습니다. extension 뒤에 의도를 표현하는 방법도 있고, 명확하지 않으면 그냥 합쳐도 됩니다.

torch-ray replied:
음료수량과 돈을 나누고자 하는 의도였는데, 생각해보니 별도로 구분할 필요가 없을 것 같아 통합했습니다!

godrm reviewed (changes requested):
수고하셨습니다. beveragesStockCount() 관련한 내용만이라도 수정하고 넘어가도록 하죠.

torch-ray committed:
뷰컨트롤러 피드백 수정 및 상수값 매개변수로 변경

뷰컨트롤러에서 count하던 것을 vendingmahcine으로 이동하고,
machine.beverages.countofBeverages()를 machine.stockOf로 수정

torch-ray committed:
Beverages class에 불필요한 메서드 삭제

torch-ray committed:
VendingMachine 구조체에서 불필요한 메서드 삭제

torch-ray committed:
ViewController에 불필요한 extension 통합

godrm commented:
다음 단계를 하면서 또 개선해보세요.
